src/components/views/login/ServerConfig.js

```diff
@@ -29,8 +29,10 @@ module.exports = React.createClass({
     propTypes: {
         onHsUrlChanged: React.PropTypes.func,
         onIsUrlChanged: React.PropTypes.func,
-        defaultHsUrl: React.PropTypes.string,
-        defaultIsUrl: React.PropTypes.string,
+        initialHsUrl: React.PropTypes.string, // whatever the current value is when we create the component
+        initialIsUrl: React.PropTypes.string, // whatever the current value is when we create the component
+        defaultHsUrl: React.PropTypes.string, // e.g. https://matrix.org
+        defaultIsUrl: React.PropTypes.string, // e.g. https://vector.im
```

Comment:
As your comment implies, the original names of the fields were misleading.

> Before this patch, the "default" URLs were the initial values which populated the input fields.

Yup, this is why they have been renamed as 'initial' URLs

> with "default" URLs now indicating what the placeholder text should be

Yup, the placeholder showing what the 'default' value for this field will be if no explicit value is specified - typically the value taken from config.json.

> Except this isn't completely true in cases where you modify the input field such that onXXXChanged fires in which case the default URL will be used if the one they entered is empty.

When is this not true?  If the input field is blank, one should surely use the default value, rather than re-animating a custom 'initial' value that happens to be hanging around from previous customisation.  If the user wanted to keep a custom 'initial' value, they should have left it there rather than clearing the field.

> We also seemingly use the defaults, overriding the explicit value in the input field, if we untick the "show advanced options" checkbox. This smells like bad UX to me as I don't think it is in any way clear than unticking that box will clobber your URLs.

This sounds subjective, but from my perspective it's abundantly clear that unchecking a checkbox labelled "Use custom server options" will stop you from using custom server options...

In terms of the underlying UX motivation here, it's:
 * Differentiate between default URLs and user-specified custom URLs
 * Let users remember custom URLs in local storage to avoid having to either bodge their config.json or repeatedly re-enter the 'right' URLs
 * Let users return easily to the correct defaults
 * Ensure that when custom URLs are specified, the "use custom server options" UI is always correctly displayed and visible.  And not bother showing it if they're using the defaults.

Does this make any more sense? @kegsay, PTAL...